### PR TITLE
Add multiple leases using metadata-templates

### PR DESCRIPTION
**Pull request by ochiengolanga**
## What does this change?

This PR replaces the ability to define a single lease (#3867) with the ability to define multiple leases in a metadata template and applying the configured "leases" to an image when the template is applied.

This is a follow-up PR to:

- #3664
- #3772 
- #3867 
- #4003

### Adding multiple leases

Metadata template with multiple add leases

### Replacing lease(s) with multiple leases

Metadata template with multiple replace leases

## How should a reviewer test this change?

Add the test metadata template configuration (below) to Kahuna's configuration.

```hocon
metadata.templates = [
  {
    templateName: "Template A w/ add"
    templateLeases: {
      replace: false
      leases: [
        {
          leaseType: "allow-use"
          notes: "Sample allow use note"
          duration: 1d
        }
        {
          leaseType: "deny-syndication"
          notes: "Sample deny syndication note"
        }
      ]
    }
  },
  {
    templateName: "Template B w/ replace"
    templateLeases: {
      replace: true
      leases: [
        {
          leaseType: "allow-use"
          notes: "Sample allow use note"
          duration: 1d
        }
        {
          leaseType: "deny-syndication"
          notes: "Sample deny syndication note"
        }
      ]
    }
  }
]
```

## How can success be measured?

Adding the above configuration to kahuna should:

- Render "Template select" component on the "My uploads" page, info panel on search images view page and Image view page as shown in the attached screenshots.
- Allow a user to select the metadata template "Template A w/add" or "Template B w/replace" that highlights multiple staged leases that have been defined in the template to add to or replace existing lease(s).
- On clicking "Apply", the staged leases should be applied to the image.

## Who should look at this?

@guardian/digital-cms

## Tested? Documented?
- [X] locally by committer
- [ ] locally by Guardian reviewer
- [x] on the Guardian's TEST environment
- [ ] relevant documentation added or amended (if needed)

### CODE REVIEW

File: kahuna/public/js/leases/leases.html [hunk 1]
```diff
@@ -13,7 +13,7 @@
        <span ng-show="ctrl.adding">Updating...</span>
     </span>
 </button>
-<span ng-if="ctrl.withBatch" class="leases__apply-all">
+<span ng-if="ctrl.withBatch && !ctrl.leasesUpdatedByTemplate" class="leases__apply-all">
```
[the remark this answers is not in the document]
Comment (ochiengolanga) [reply kahuna/public/js/leases/leases.html:16]: This suggestion makes sense. I've pushed a commit to make the hide "apply-to-all" button consistent for fields that could be updated by metadata-templates and are in "preview" mode before application of a metadata template. I'm happy for you to have a look.

**Comment by paperboyo** [comment]
Great testing, @andrew-nowak! I would only say that yes, we are not using them now, but maybe we will after https://github.com/guardian/grid/discussions/4044.

So worth making them work well. Second bug should be fixed. First one (replacement; not adding) is interesting. What happens to templates that operate on arrays (eg. keywords)? Maybe teplate config should be able to specify `replace` and `add`? (if not `remove`)?

File: kahuna/public/js/metadata-templates/metadata-templates.js [hunk 2]
```diff
@@ -165,9 +165,9 @@ metadataTemplates.controller('MetadataTemplatesCtrl', [
       }
     })
     .then(() => {
-      if (ctrl.metadataTemplate.lease) {
-        const lease = toLease(ctrl.metadataTemplate.lease);
-        return leaseService.batchAdd(lease, [ctrl.image]);
+      if (ctrl.metadataTemplate.leases) {
+        const leases = ctrl.metadataTemplate.leases.map(lease => toLease(lease));
+        return leaseService.replace(ctrl.image, leases);
```
[the remark this answers is not in the document]
Comment (ochiengolanga) [reply kahuna/public/js/metadata-templates/metadata-templates.js]: Hi @andrew-nowak , great question. Initially, we were looking at adding leases to newly uploaded images that didn't have any existing lease(s). Looking into your comment on the unexpected behaviour, there could be a case where a template could be applied to an image that already has a lease(s) and depending on the template, the intention could be to add or to replace any existing leases.

I've made a commit that now supports both intentions by introducing the following:
- Added a new configuration `templateLeases.replace` to define where a template should add or replace leases on an image when applied.
- Added a new API endpoint `POST /leases/media/:id` that can be used to add multiple leases.
- Modified the replace leases API endpoint `/leases/media/:id` method from `POST` to `PUT` 

Both are ready for test with the sample template configuration in the description.

**Comment by ochiengolanga** [comment]
> Great testing, @andrew-nowak! I would only say that yes, we are not using them now, but maybe we will after #4044.
> 
> So worth making them work well. Second bug should be fixed. First one (replacement; not adding) is interesting. What happens to templates that operate on arrays (eg. keywords)? Maybe teplate config should be able to specify `replace` and `add`? (if not `remove`)?

@paperboyo Great suggestion. The proposed configuration is now part of the template config.

**Review by rebecca-thompson — COMMENTED** [review]
I have reviewed the code and it looks good to me. I noticed a couple of visual issues, but as this is a BBC feature I'm happy for them to be non-blocking and approve this pr if you find they are not critical to this feature being released.

The first is the positioning of the buttons. It looks like they could use some top padding. The second is a flickering that happens when the new leases are applied. When the apply button is clicked, the original leases display first and are then replaced by the new leases. Probably best explained by the short clip below.

Either way, let me know and I will await changes, or approve and merge.

https://github.com/guardian/grid/assets/33927854/dacd8913-c78b-48c1-9a5e-8d91c2742f11

**Comment by ochiengolanga** [comment]
Thanks for the feedback @rebecca-thompson. I'll take a look at the flickering and add spacing on the buttons.

**Comment by ochiengolanga** [comment]
> I have reviewed the code and it looks good to me. I noticed a couple of visual issues, but as this is a BBC feature I'm happy for them to be non-blocking and approve this pr if you find they are not critical to this feature being released.
> 
> The first is the positioning of the buttons. It looks like they could use some top padding. The second is a flickering that happens when the new leases are applied. When the apply button is clicked, the original leases display first and are then replaced by the new leases. Probably best explained by the short clip below.
> 
> Either way, let me know and I will await changes, or approve and merge.
> 
>  leases.mov

Hi @rebecca-thompson , both visual issues should be fixed by the recent commit. Feel free to give it a go.

**Review by rebecca-thompson — APPROVED** [review]

**Pull request merged**